geopandas/sindex.py

```diff
@@ -1,25 +1,235 @@
-from geopandas import base
+from warnings import warn
+from collections import namedtuple
 
-if base.HAS_SINDEX:
-    from rtree.index import Index as RTreeIndex
+from pandas import notnull
+import numpy as np
 
+from . import _compat as compat
 
-class SpatialIndex(RTreeIndex):
+
+def has_sindex():
+    """
+    Dynamically checks for ability to generate spatial index.
+    """
+    return get_sindex_class() is not None
+
+
+def get_sindex_class():
     """
-    A simple wrapper around rtree's RTree Index
+    Dynamically chooses a spatial indexing backend.
+    Required to comply with _compat.USE_PYGEOS.
+    The order of preference goes PyGeos > RTree > None.
     """
+    if compat.USE_PYGEOS:
+        return PyGEOSSTRTreeIndex
+    elif compat.HAS_RTREE:
+        return RTreeIndex
+    else:
+        warn("Cannot generate spatial index: Missing package `rtree` or `pygeos`.")
+        return None
+
+
+if compat.HAS_RTREE:
+
+    from rtree.index import Index as ToblerityRTreeIndex  # noqa
+    from rtree.core import RTreeError  # noqa
+    from shapely.geometry import box
+
+    class RTreeIndex(ToblerityRTreeIndex):
+        """
+        A simple wrapper around rtree's RTree Index
+        """
+
+        def __init__(self, geometry):
+            stream = (
+                (i, item.bounds, idx)
+                for i, (idx, item) in enumerate(geometry.iteritems())
+                if notnull(item) and not item.is_empty
+            )
+            try:
+                super().__init__(stream)
+            except RTreeError:
+                # What we really want here is an empty generator error, or
+                # for the bulk loader to log that the generator was empty
+                # and move on.
+                # See https://github.com/Toblerity/rtree/issues/20.
+                super().__init__()
+
+            # store reference to geometries for predicate queries
+            self._geometries = geometry.geometry
```

Comment:
might want to store an array instead of the Series (because indexing into it as you do below might go wrong with a non-default index)

Reply:
Makes sense.

Is `self._geometries = geometry.geometry.values`? I still get a bit confused as to what is created/converted on the fly and what already lives in memory. Obviously here we want to store a reference to the existing data in memory, not a reference to temporary data, which would increase memory usage.

Reply:
`geometry.geometry.values` should be fine, if the input `geometry` is a GeoSeries or GeoDataFrame